File: edge-gateway-ui/src/api/dataPointTemplate.ts
import request from './request'

export interface DataPointTemplateItem {
  id: number
  name: string
  description?: string | null
  protocol: string
  pointCount: number
  createdAt: string
  updatedAt: string
}

export interface DataPointTemplatePoint {
  name: string
  tagSuffix: string
  description?: string | null
  address: string
  dataType: number
  unit?: string | null
  modbusSlaveId?: number | null
  modbusFunctionCode?: number | null
  modbusByteOrder?: number | null
  registerLength: number
  modbusBitIndex?: number | null
  isEnabled: boolean
  isControllable: boolean
}

export interface DataPointTemplateDetail extends DataPointTemplateItem {
  points: DataPointTemplatePoint[]
}

export const getDataPointTemplates = (protocol?: number) =>
  request.get('/datapoint-templates', { params: protocol === undefined ? undefined : { protocol } })

export const createDataPointTemplateFromDevice = (deviceId: number, data: { name: string; description?: string }) =>
  request.post(`/datapoint-templates/from-device/${deviceId}`, data)

export const applyDataPointTemplate = (deviceId: number, templateId: number, overwriteExisting = false) =>
  request.post(`/devices/${deviceId}/datapoints/templates/${templateId}/apply`, { overwriteExisting })

export const deleteDataPointTemplate = (templateId: number) =>
  request.delete(`/datapoint-templates/${templateId}`)

export const getDataPointTemplatesPaged = (params: { page?: number; pageSize?: number; search?: string; protocol?: number }) =>
  request.get('/datapoint-templates/paged', { params })

export const getDataPointTemplate = (templateId: number) =>
  request.get(`/datapoint-templates/${templateId}`)

export const updateDataPointTemplate = (templateId: number, data: { name: string; description?: string; points: DataPointTemplatePoint[] }) =>
  request.put(`/datapoint-templates/${templateId}`, data)

File: EdgeGateway.Domain/Entities/DataPointTemplate.cs
using EdgeGateway.Domain.Enums;

namespace EdgeGateway.Domain.Entities;

/// <summary>可复用的数据点模板。</summary>
public class DataPointTemplate
{
    public int Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string? Description { get; set; }
    public CollectionProtocol Protocol { get; set; }
    public string PointsJson { get; set; } = "[]";
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

/// <summary>模板中的数据点配置，不包含设备 ID 和运行时数据。</summary>
public sealed class DataPointTemplateItem
{
    public string Name { get; set; } = string.Empty;
    public string TagSuffix { get; set; } = string.Empty;
    public string? Description { get; set; }
    public string Address { get; set; } = string.Empty;
    public DataValueType DataType { get; set; }
    public string? Unit { get; set; }
    public byte? ModbusSlaveId { get; set; }
    public int? ModbusFunctionCode { get; set; }
    public ModbusByteOrder? ModbusByteOrder { get; set; }
    public byte RegisterLength { get; set; } = 1;
    public byte? ModbusBitIndex { get; set; }
    public bool IsEnabled { get; set; } = true;
    public bool IsControllable { get; set; }
}
